File: receiver-go/protocol.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math"
	"strconv"
)

const (
	protocolVersion = 1
	maxFrameBits    = 1_000_000
)

type Request struct {
	ProtocolVersion uint64
	RequestID       string
	Algorithm       string
	SourceOctets    uint64
	FrameBits       string
	Noise           Noise
}

type Noise struct {
	ProbabilityNumerator   uint64
	ProbabilityDenominator uint64
	Seed                   uint64
	FlippedBits            uint64
}

type wireRequest struct {
	ProtocolVersion *json.RawMessage `json:"protocol_version"`
	RequestID       *string          `json:"request_id"`
	Algorithm       *string          `json:"algorithm"`
	SourceOctets    *json.RawMessage `json:"source_octets"`
	FrameBits       *string          `json:"frame_bits"`
	Noise           *wireNoise       `json:"noise"`
}

type wireNoise struct {
	ProbabilityNumerator   *json.RawMessage `json:"probability_numerator"`
	ProbabilityDenominator *json.RawMessage `json:"probability_denominator"`
	Seed                   *json.RawMessage `json:"seed"`
	FlippedBits            *json.RawMessage `json:"flipped_bits"`
}

type Response struct {
	ProtocolVersion int            `json:"protocol_version"`
	RequestID       *string        `json:"request_id"`
	Status          string         `json:"status"`
	Message         *string        `json:"message"`
	Error           *ResponseError `json:"error"`
	Metrics         *Metrics       `json:"metrics"`
}

type ResponseError struct {
	Code   string `json:"code"`
	Detail string `json:"detail"`
}

type Metrics struct {
	ReceivedBits        uint64 `json:"received_bits"`
	SourceBits          uint64 `json:"source_bits"`
	RedundancyBits      uint64 `json:"redundancy_bits"`
	ReportedFlippedBits uint64 `json:"reported_flipped_bits"`
	DetectedUnits       uint64 `json:"detected_units"`
	CorrectedBits       uint64 `json:"corrected_bits"`
	UncorrectableUnits  uint64 `json:"uncorrectable_units"`
}

type protocolError struct {
	code   string
	detail string
}

func processLine(line []byte) Response {
	request, requestID, err := parseRequest(line)
	if err != nil {
		return invalidResponse(requestID, err.code, err.detail)
	}
	return decodeRequest(request)
}

func parseRequest(line []byte) (Request, *string, *protocolError) {
	if len(line) == 0 {
		return Request{}, nil, &protocolError{"invalid_json", "request line must contain one JSON object"}
	}
	if err := validateJSONShape(line); err != nil {
		code := "invalid_json"
		if errors.Is(err, errDuplicateKey) {
			code = "invalid_schema"
		}
		return Request{}, nil, &protocolError{code, err.Error()}
	}

	requestID := recoverRequestID(line)
	decoder := json.NewDecoder(bytes.NewReader(line))
	decoder.UseNumber()
	decoder.DisallowUnknownFields()
	var wire wireRequest
	if err := decoder.Decode(&wire); err != nil {
		return Request{}, requestID, &protocolError{"invalid_schema", "request does not match the exact schema"}
	}
	if wire.ProtocolVersion == nil || wire.RequestID == nil || wire.Algorithm == nil ||
		wire.SourceOctets == nil || wire.FrameBits == nil || wire.Noise == nil ||
		wire.Noise.ProbabilityNumerator == nil || wire.Noise.ProbabilityDenominator == nil ||
		wire.Noise.Seed == nil || wire.Noise.FlippedBits == nil {
		return Request{}, requestID, &protocolError{"invalid_schema", "all request and noise fields are required and non-null"}
	}

	version, ok := unsignedInteger(*wire.ProtocolVersion)
	if !ok {
		return Request{}, requestID, &protocolError{"invalid_schema", "protocol_version must be a non-negative JSON integer"}
	}
	sourceOctets, ok := unsignedInteger(*wire.SourceOctets)
	if !ok {
		return Request{}, requestID, &protocolError{"invalid_schema", "source_octets must be a non-negative JSON integer"}
	}
	numerator, numeratorOK := unsignedInteger(*wire.Noise.ProbabilityNumerator)
	denominator, denominatorOK := unsignedInteger(*wire.Noise.ProbabilityDenominator)
	seed, seedOK := unsignedInteger(*wire.Noise.Seed)
	flippedBits, flippedOK := unsignedInteger(*wire.Noise.FlippedBits)
	if !numeratorOK || !denominatorOK || !seedOK || !flippedOK {
		return Request{}, requestID, &protocolError{"invalid_schema", "noise values must be non-negative JSON integers"}
	}

	request := Request{
		ProtocolVersion: version,
		RequestID:       *wire.RequestID,
		Algorithm:       *wire.Algorithm,
		SourceOctets:    sourceOctets,
		FrameBits:       *wire.FrameBits,
		Noise: Noise{
			ProbabilityNumerator:   numerator,
			ProbabilityDenominator: denominator,
			Seed:                   seed,
			FlippedBits:            flippedBits,
		},
	}
	if version != protocolVersion {
		return Request{}, requestID, &protocolError{"unsupported_version", "protocol_version must be 1"}
	}
	if !validRequestID(request.RequestID) {
		return Request{}, nil, &protocolError{"invalid_schema", "request_id must match [A-Za-z0-9._-]{1,64}"}
	}
	requestID = &request.RequestID
	if request.Algorithm != "crc32-iso-hdlc" && request.Algorithm != "hamming-secded-13-8" {
		return Request{}, requestID, &protocolError{"invalid_schema", "algorithm is not supported"}
	}
	if request.SourceOctets == 0 {
		return Request{}, requestID, &protocolError{"invalid_schema", "source_octets must be at least 1"}
	}
	if len(request.FrameBits) > maxFrameBits || !onlyBits(request.FrameBits) {
		return Request{}, requestID, &protocolError{"invalid_schema", "frame_bits must contain at most 1,000,000 binary digits"}
	}
	if denominator == 0 || denominator > 1_000_000_000 || numerator > denominator {
		return Request{}, requestID, &protocolError{"invalid_schema", "noise probability is outside the allowed range"}
	}
	if flippedBits > uint64(len(request.FrameBits)) {
		return Request{}, requestID, &protocolError{"invalid_schema", "noise.flipped_bits exceeds frame length"}
	}
	if !validFrameLength(request) {
		return Request{}, requestID, &protocolError{"invalid_frame_length", "frame length does not match algorithm and source_octets"}
	}
	return request, requestID, nil
}

func decodeRequest(request Request) Response {
	metrics := Metrics{
		ReceivedBits:        uint64(len(request.FrameBits)),
		SourceBits:          request.SourceOctets * 8,
		ReportedFlippedBits: request.Noise.FlippedBits,
	}
	var result decodeResult
	if request.Algorithm == "crc32-iso-hdlc" {
		metrics.RedundancyBits = 32
		result = decodeCRC(request.FrameBits)
	} else {
		metrics.RedundancyBits = request.SourceOctets * 5
		result = decodeHamming(request.FrameBits)
	}
	metrics.DetectedUnits = result.detectedUnits
	metrics.CorrectedBits = result.correctedBits
	metrics.UncorrectableUnits = result.uncorrectableUnits

	if result.errorCode != "" {
		return failureResponse(request.RequestID, result.errorCode, result.detail, metrics)
	}
	if !printableASCII(result.payload) {
		return failureResponse(request.RequestID, "invalid_ascii_payload", "recovered payload is not printable ASCII", metrics)
	}
	message := string(result.payload)
	status := "ok"
	if result.correctedBits > 0 {
		status = "corrected"
	}
	return Response{protocolVersion, &request.RequestID, status, &message, nil, &metrics}
}

func invalidResponse(requestID *string, code, detail string) Response {
	return Response{protocolVersion, requestID, "invalid_request", nil, &ResponseError{code, detail}, nil}
}

func failureResponse(requestID, code, detail string, metrics Metrics) Response {
	return Response{protocolVersion, &requestID, "detected_uncorrectable", nil, &ResponseError{code, detail}, &metrics}
}

func validRequestID(value string) bool {
	if len(value) < 1 || len(value) > 64 {
		return false
	}
	for _, character := range []byte(value) {
		if !(character >= 'A' && character <= 'Z') && !(character >= 'a' && character <= 'z') &&
			!(character >= '0' && character <= '9') && character != '.' && character != '_' && character != '-' {
			return false
		}
	}
	return true
}

func onlyBits(value string) bool {
	if value == "" {
		return false
	}
	for _, character := range []byte(value) {
		if character != '0' && character != '1' {
			return false
		}
	}
	return true
}

func printableASCII(payload []byte) bool {
	if len(payload) == 0 {
		return false
	}
	for _, value := range payload {
		if value < 0x20 || value > 0x7e {
			return false
		}
	}
	return true
}

func validFrameLength(request Request) bool {
	var expected uint64
	if request.Algorithm == "crc32-iso-hdlc" {
		if request.SourceOctets > (math.MaxUint64-32)/8 {
			return false
		}
		expected = request.SourceOctets*8 + 32
	} else {
		if request.SourceOctets > math.MaxUint64/13 {
			return false
		}
		expected = request.SourceOctets * 13
	}
	return expected == uint64(len(request.FrameBits))
}

func unsignedInteger(raw json.RawMessage) (uint64, bool) {
	text := string(raw)
	if text == "" || text[0] == '-' {
		return 0, false
	}
	for _, character := range text {
		if character < '0' || character > '9' {
			return 0, false
		}
	}
	value, err := strconv.ParseUint(text, 10, 64)
	return value, err == nil
}

func recoverRequestID(line []byte) *string {
	var object map[string]json.RawMessage
	if json.Unmarshal(line, &object) != nil {
		return nil
	}
	var requestID string
	if raw, exists := object["request_id"]; !exists || json.Unmarshal(raw, &requestID) != nil || !validRequestID(requestID) {
		return nil
	}
	return &requestID
}

var errDuplicateKey = errors.New("duplicate JSON object key")

func validateJSONShape(line []byte) error {
	decoder := json.NewDecoder(bytes.NewReader(line))
	decoder.UseNumber()
	token, err := decoder.Token()
	if err != nil {
		return fmt.Errorf("request is not valid JSON: %w", err)
	}
	if delimiter, ok := token.(json.Delim); !ok || delimiter != '{' {
		return errors.New("request JSON value must be an object")
	}
	if err := consumeObject(decoder); err != nil {
		return err
	}
	if _, err := decoder.Token(); !errors.Is(err, io.EOF) {
		if err == nil {
			return errors.New("request line contains multiple JSON values")
		}
		return fmt.Errorf("request is not valid JSON: %w", err)
	}
	return nil
}

func consumeObject(decoder *json.Decoder) error {
	keys := make(map[string]struct{})
	for decoder.More() {
		token, err := decoder.Token()
		if err != nil {
			return fmt.Errorf("request is not valid JSON: %w", err)
		}
		key, ok := token.(string)
		if !ok {
			return errors.New("request object key must be a string")
		}
		if _, exists := keys[key]; exists {
			return fmt.Errorf("%w: %q", errDuplicateKey, key)
		}
		keys[key] = struct{}{}
		if err := consumeValue(decoder); err != nil {
			return err
		}
	}
	if _, err := decoder.Token(); err != nil {
		return fmt.Errorf("request is not valid JSON: %w", err)
	}
	return nil
}

func consumeValue(decoder *json.Decoder) error {
	token, err := decoder.Token()
	if err != nil {
		return fmt.Errorf("request is not valid JSON: %w", err)
	}
	delimiter, ok := token.(json.Delim)
	if !ok {
		return nil
	}
	switch delimiter {
	case '{':
		return consumeObject(decoder)
	case '[':
		for decoder.More() {
			if err := consumeValue(decoder); err != nil {
				return err
			}
		}
		_, err := decoder.Token()
		return err
	default:
		return errors.New("unexpected JSON delimiter")
	}
}

File: sender-rust/src/protocol.rs
use serde::{Deserialize, Serialize};

use crate::{ascii, crc, hamming, noise};

pub const PROTOCOL_VERSION: u8 = 1;
pub const MAX_FRAME_BITS: usize = 1_000_000;
pub const MAX_LINE_BYTES: usize = 1_048_576;

#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Eq, Serialize)]
pub enum Algorithm {
    #[serde(rename = "crc32-iso-hdlc")]
    Crc32IsoHdlc,
    #[serde(rename = "hamming-secded-13-8")]
    HammingSecded13_8,
}

impl Algorithm {
    pub fn parse(value: &str) -> Result<Self, String> {
        match value {
            "crc" | "crc32-iso-hdlc" => Ok(Self::Crc32IsoHdlc),
            "hamming" | "hamming-secded-13-8" => Ok(Self::HammingSecded13_8),
            _ => Err(format!(
                "unsupported algorithm '{value}'; use crc32-iso-hdlc or hamming-secded-13-8"
            )),
        }
    }
}

#[derive(Debug, Deserialize, PartialEq, Eq, Serialize)]
#[serde(deny_unknown_fields)]
pub struct NoiseMetadata {
    pub probability_numerator: u64,
    pub probability_denominator: u64,
    pub seed: u64,
    pub flipped_bits: u64,
}

#[derive(Debug, Deserialize, PartialEq, Eq, Serialize)]
#[serde(deny_unknown_fields)]
pub struct Request {
    pub protocol_version: u8,
    pub request_id: String,
    pub algorithm: Algorithm,
    pub source_octets: usize,
    pub frame_bits: String,
    pub noise: NoiseMetadata,
    #[serde(skip)]
    original_message: String,
}

pub fn build_request(
    request_id: &str,
    algorithm: Algorithm,
    message: &str,
    probability_numerator: u64,
    probability_denominator: u64,
    seed: u64,
) -> Result<Request, String> {
    validate_request_id(request_id)?;
    ascii::validate_printable_ascii(message)?;
    noise::validate_probability(probability_numerator, probability_denominator)?;

    let clean_frame = match algorithm {
        Algorithm::Crc32IsoHdlc => crc::encode_frame(message.as_bytes()),
        Algorithm::HammingSecded13_8 => hamming::encode_frame(message.as_bytes()),
    };
    if clean_frame.len() > MAX_FRAME_BITS {
        return Err(format!(
            "encoded frame has {} bits; maximum is {MAX_FRAME_BITS}",
            clean_frame.len()
        ));
    }

    let noisy = noise::apply_noise(
        &clean_frame,
        probability_numerator,
        probability_denominator,
        seed,
    )?;

    Ok(Request {
        protocol_version: PROTOCOL_VERSION,
        request_id: request_id.to_owned(),
        algorithm,
        source_octets: message.len(),
        frame_bits: noisy.frame_bits,
        noise: NoiseMetadata {
            probability_numerator,
            probability_denominator,
            seed,
            flipped_bits: noisy.flipped_bits,
        },
        original_message: message.to_owned(),
    })
}

pub fn validate_request_id(request_id: &str) -> Result<(), String> {
    if !(1..=64).contains(&request_id.len())
        || !request_id
            .bytes()
            .all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'.' | b'_' | b'-'))
    {
        return Err(
            "request ID must contain 1-64 ASCII letters, digits, dots, underscores, or hyphens"
                .to_owned(),
        );
    }
    Ok(())
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Response {
    pub protocol_version: u8,
    #[serde(deserialize_with = "required_nullable")]
    pub request_id: Option<String>,
    pub status: String,
    #[serde(deserialize_with = "required_nullable")]
    pub message: Option<String>,
    #[serde(deserialize_with = "required_nullable")]
    pub error: Option<ResponseError>,
    #[serde(deserialize_with = "required_nullable")]
    pub metrics: Option<Metrics>,
}

fn required_nullable<'de, D, T>(deserializer: D) -> Result<Option<T>, D::Error>
where
    D: serde::Deserializer<'de>,
    T: Deserialize<'de>,
{
    Option::<T>::deserialize(deserializer)
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ResponseError {
    pub code: String,
    pub detail: String,
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Metrics {
    pub received_bits: u64,
    pub source_bits: u64,
    pub redundancy_bits: u64,
    pub reported_flipped_bits: u64,
    pub detected_units: u64,
    pub corrected_bits: u64,
    pub uncorrectable_units: u64,
}

impl Response {
    pub fn validate(&self, request: &Request) -> Result<(), String> {
        if self.protocol_version != request.protocol_version {
            return Err(format!(
                "response protocol version is {}, expected {}",
                self.protocol_version, request.protocol_version
            ));
        }
        if let Some(request_id) = &self.request_id
            && request_id != &request.request_id
        {
            return Err(format!(
                "response request ID '{request_id}' does not match '{}'",
                request.request_id
            ));
        }

        if request.noise.flipped_bits == 0
            && (self.status != "ok" || self.message.as_deref() != Some(&request.original_message))
        {
            return Err(
                "zero-noise response must be successful and preserve the original message"
                    .to_owned(),
            );
        }

        match self.status.as_str() {
            "ok" | "corrected" => {
                if self.request_id.is_none()
                    || self.message.is_none()
                    || self.error.is_some()
                    || self.metrics.is_none()
                {
                    return Err(
                        "success response has inconsistent message/error/metrics fields".to_owned(),
                    );
                }
                let message = self.message.as_deref().unwrap_or_default();
                ascii::validate_printable_ascii(message)
                    .map_err(|error| format!("response message is invalid: {error}"))?;
                if message.len() != request.source_octets {
                    return Err(format!(
                        "response message has {} octets, expected {}",
                        message.len(),
                        request.source_octets
                    ));
                }
            }
            "detected_uncorrectable" => {
                if self.request_id.is_none()
                    || self.message.is_some()
                    || self.error.is_none()
                    || self.metrics.is_none()
                {
                    return Err(
                        "uncorrectable response has inconsistent message/error/metrics fields"
                            .to_owned(),
                    );
                }
            }
            "invalid_request" => {
                if self.message.is_some() || self.error.is_none() || self.metrics.is_some() {
                    return Err(
                        "invalid-request response has inconsistent message/error/metrics fields"
                            .to_owned(),
                    );
                }
            }
            status => return Err(format!("response contains unknown status '{status}'")),
        }

        let error_code = self.error.as_ref().map(|error| error.code.as_str());
        let valid_error = match self.status.as_str() {
            "ok" | "corrected" => error_code.is_none(),
            "invalid_request" => matches!(
                error_code,
                Some(
                    "invalid_json"
                        | "line_too_long"
                        | "unsupported_version"
                        | "invalid_schema"
                        | "invalid_frame_length"
                )
            ),
            "detected_uncorrectable" => match request.algorithm {
                Algorithm::Crc32IsoHdlc => matches!(
                    error_code,
                    Some("integrity_check_failed" | "invalid_ascii_payload")
                ),
                Algorithm::HammingSecded13_8 => matches!(
                    error_code,
                    Some("uncorrectable_error" | "invalid_ascii_payload")
                ),
            },
            _ => false,
        };
        if !valid_error {
            return Err(format!(
                "response status '{}' is inconsistent with error code {:?}",
                self.status, error_code
            ));
        }

        if let Some(metrics) = &self.metrics {
            validate_metrics(metrics, request, &self.status, error_code)?;
        }

        Ok(())
    }
}

fn validate_metrics(
    metrics: &Metrics,
    request: &Request,
    status: &str,
    error_code: Option<&str>,
) -> Result<(), String> {
    let received_bits = u64::try_from(request.frame_bits.len())
        .map_err(|_| "request frame length cannot be represented as u64".to_owned())?;
    let source_octets = u64::try_from(request.source_octets)
        .map_err(|_| "request source length cannot be represented as u64".to_owned())?;
    let source_bits = source_octets
        .checked_mul(8)
        .ok_or_else(|| "request source bit count overflowed".to_owned())?;
    let redundancy_bits = match request.algorithm {
        Algorithm::Crc32IsoHdlc => 32,
        Algorithm::HammingSecded13_8 => source_octets
            .checked_mul(5)
            .ok_or_else(|| "request redundancy bit count overflowed".to_owned())?,
    };

    let expected = [
        ("received_bits", metrics.received_bits, received_bits),
        ("source_bits", metrics.source_bits, source_bits),
        ("redundancy_bits", metrics.redundancy_bits, redundancy_bits),
        (
            "reported_flipped_bits",
            metrics.reported_flipped_bits,
            request.noise.flipped_bits,
        ),
    ];
    for (name, actual, expected) in expected {
        if actual != expected {
            return Err(format!(
                "response metric {name} is {actual}, expected {expected}"
            ));
        }
    }

    let counters_are = |detected, corrected, uncorrectable| {
        metrics.detected_units == detected
            && metrics.corrected_bits == corrected
            && metrics.uncorrectable_units == uncorrectable
    };
    let counters_valid = match (request.algorithm, status, error_code) {
        (_, "ok", None) => counters_are(0, 0, 0),
        (Algorithm::Crc32IsoHdlc, "detected_uncorrectable", Some("integrity_check_failed")) => {
            counters_are(1, 0, 1)
        }
        (Algorithm::Crc32IsoHdlc, "detected_uncorrectable", Some("invalid_ascii_payload")) => {
            counters_are(0, 0, 0)
        }
        (Algorithm::HammingSecded13_8, "corrected", None) => {
            metrics.corrected_bits > 0
                && metrics.detected_units == metrics.corrected_bits
                && metrics.uncorrectable_units == 0
                && metrics.detected_units <= source_octets
        }
        (Algorithm::HammingSecded13_8, "detected_uncorrectable", Some("uncorrectable_error")) => {
            metrics.uncorrectable_units > 0
                && metrics
                    .corrected_bits
                    .checked_add(metrics.uncorrectable_units)
                    == Some(metrics.detected_units)
                && metrics.detected_units <= source_octets
        }
        (Algorithm::HammingSecded13_8, "detected_uncorrectable", Some("invalid_ascii_payload")) => {
            metrics.uncorrectable_units == 0
                && metrics.detected_units == metrics.corrected_bits
                && metrics.detected_units <= source_octets
        }
        _ => false,
    };
    if !counters_valid {
        return Err(format!(
            "response counters are inconsistent with algorithm {:?}, status '{status}', and error code {error_code:?}",
            request.algorithm
        ));
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn request(algorithm: Algorithm, flipped_bits: u64) -> Request {
        Request {
            protocol_version: PROTOCOL_VERSION,
            request_id: "demo-1".to_owned(),
            algorithm,
            source_octets: 1,
            frame_bits: match algorithm {
                Algorithm::Crc32IsoHdlc => "0".repeat(40),
                Algorithm::HammingSecded13_8 => "0".repeat(13),
            },
            noise: NoiseMetadata {
                probability_numerator: 0,
                probability_denominator: 1,
                seed: 0,
                flipped_bits,
            },
            original_message: "A".to_owned(),
        }
    }

    fn response(json: &str) -> Response {
        serde_json::from_str(json).expect("valid response schema")
    }

    #[test]
    fn accepts_a_clean_zero_noise_response() {
        let response: Response = serde_json::from_str(
            r#"{"protocol_version":1,"request_id":"demo-1","status":"ok","message":"A","error":null,"metrics":{"received_bits":13,"source_bits":8,"redundancy_bits":5,"reported_flipped_bits":0,"detected_units":0,"corrected_bits":0,"uncorrectable_units":0}}"#,
        )
        .expect("valid response schema");
        assert_eq!(
            response.validate(&request(Algorithm::HammingSecded13_8, 0)),
            Ok(())
        );
    }

    #[test]
    fn accepts_clean_crc_zero_noise_response() {
        let response = response(
            r#"{"protocol_version":1,"request_id":"demo-1","status":"ok","message":"A","error":null,"metrics":{"received_bits":40,"source_bits":8,"redundancy_bits":32,"reported_flipped_bits":0,"detected_units":0,"corrected_bits":0,"uncorrectable_units":0}}"#,
        );

        assert_eq!(
            response.validate(&request(Algorithm::Crc32IsoHdlc, 0)),
            Ok(())
        );
    }

    #[test]
    fn rejects_hamming_correction_with_zero_noise() {
        let response = response(
            r#"{"protocol_version":1,"request_id":"demo-1","status":"corrected","message":"A","error":null,"metrics":{"received_bits":13,"source_bits":8,"redundancy_bits":5,"reported_flipped_bits":0,"detected_units":1,"corrected_bits":1,"uncorrectable_units":0}}"#,
        );

        assert!(
            response
                .validate(&request(Algorithm::HammingSecded13_8, 0))
                .is_err()
        );
    }

    #[test]
    fn rejects_crc_integrity_failure_with_zero_noise() {
        let response = response(
            r#"{"protocol_version":1,"request_id":"demo-1","status":"detected_uncorrectable","message":null,"error":{"code":"integrity_check_failed","detail":"bad CRC"},"metrics":{"received_bits":40,"source_bits":8,"redundancy_bits":32,"reported_flipped_bits":0,"detected_units":1,"corrected_bits":0,"uncorrectable_units":1}}"#,
        );

        assert!(
            response
                .validate(&request(Algorithm::Crc32IsoHdlc, 0))
                .is_err()
        );
    }

    #[test]
    fn rejects_changed_message_with_zero_noise() {
        let response = response(
            r#"{"protocol_version":1,"request_id":"demo-1","status":"ok","message":"B","error":null,"metrics":{"received_bits":13,"source_bits":8,"redundancy_bits":5,"reported_flipped_bits":0,"detected_units":0,"corrected_bits":0,"uncorrectable_units":0}}"#,
        );

        assert!(
            response
                .validate(&request(Algorithm::HammingSecded13_8, 0))
                .is_err()
        );
    }

    #[test]
    fn rejects_missing_nullable_fields_and_invalid_semantics() {
        let missing_error = r#"{"protocol_version":1,"request_id":"demo-1","status":"ok","message":"A","metrics":{"received_bits":13,"source_bits":8,"redundancy_bits":5,"reported_flipped_bits":0,"detected_units":0,"corrected_bits":0,"uncorrectable_units":0}}"#;
        assert!(serde_json::from_str::<Response>(missing_error).is_err());

        let unknown_status: Response = serde_json::from_str(
            r#"{"protocol_version":1,"request_id":"demo-1","status":"maybe","message":null,"error":null,"metrics":null}"#,
        )
        .expect("valid schema");
        assert!(
            unknown_status
                .validate(&request(Algorithm::HammingSecded13_8, 0))
                .is_err()
        );
    }

    #[test]
    fn rejects_success_with_impossible_metrics() {
        let response = response(
            r#"{"protocol_version":1,"request_id":"demo-1","status":"ok","message":"A","error":null,"metrics":{"received_bits":0,"source_bits":0,"redundancy_bits":0,"reported_flipped_bits":999,"detected_units":1,"corrected_bits":1,"uncorrectable_units":1}}"#,
        );

        assert!(
            response
                .validate(&request(Algorithm::HammingSecded13_8, 2))
                .is_err()
        );
    }

    #[test]
    fn rejects_invalid_status_error_combinations() {
        let invalid_request_with_integrity_error = response(
            r#"{"protocol_version":1,"request_id":"demo-1","status":"invalid_request","message":null,"error":{"code":"integrity_check_failed","detail":"bad"},"metrics":null}"#,
        );
        assert!(
            invalid_request_with_integrity_error
                .validate(&request(Algorithm::Crc32IsoHdlc, 0))
                .is_err()
        );

        let detected_with_schema_error = response(
            r#"{"protocol_version":1,"request_id":"demo-1","status":"detected_uncorrectable","message":null,"error":{"code":"invalid_schema","detail":"bad"},"metrics":{"received_bits":40,"source_bits":8,"redundancy_bits":32,"reported_flipped_bits":0,"detected_units":1,"corrected_bits":0,"uncorrectable_units":1}}"#,
        );
        assert!(
            detected_with_schema_error
                .validate(&request(Algorithm::Crc32IsoHdlc, 0))
                .is_err()
        );
    }

    #[test]
    fn rejects_algorithm_inconsistent_status_and_counters() {
        let corrected_crc = response(
            r#"{"protocol_version":1,"request_id":"demo-1","status":"corrected","message":"A","error":null,"metrics":{"received_bits":40,"source_bits":8,"redundancy_bits":32,"reported_flipped_bits":1,"detected_units":1,"corrected_bits":1,"uncorrectable_units":0}}"#,
        );
        assert!(
            corrected_crc
                .validate(&request(Algorithm::Crc32IsoHdlc, 1))
                .is_err()
        );

        let corrected_without_correction = response(
            r#"{"protocol_version":1,"request_id":"demo-1","status":"corrected","message":"A","error":null,"metrics":{"received_bits":13,"source_bits":8,"redundancy_bits":5,"reported_flipped_bits":1,"detected_units":0,"corrected_bits":0,"uncorrectable_units":0}}"#,
        );
        assert!(
            corrected_without_correction
                .validate(&request(Algorithm::HammingSecded13_8, 1))
                .is_err()
        );
    }
}
